devtools/x/src/lint/guppy.rs

```diff
@@ -507,3 +507,167 @@ impl PackageLinter for OnlyPublishToCratesIo {
+    ("move-lang-functional-tests", "functional-tests"),
+    ("move-lang-functional-tests", "diem-types"),
+    ("move-lang-functional-tests", "datatest-stable"),
+    ("move-lang-test-utils", "datatest-stable"),
+    ("move-model", "diem-types"),
+    ("move-model", "datatest-stable"),
+    ("move-prover", "diem-types"),
+    ("move-prover", "diem-temppath"),
+    ("move-prover", "datatest-stable"),
+    ("abigen", "diem-types"),
+    ("abigen", "diem-temppath"),
+    ("abigen", "datatest-stable"),
+    ("boogie-backend", "diem-types"),
+    ("boogie-backend", "diem-temppath"),
+    ("boogie-backend", "datatest-stable"),
+    ("bytecode", "diem-types"),
+    ("bytecode", "diem-temppath"),
+    ("bytecode", "datatest-stable"),
+    ("docgen", "diem-temppath"),
+    ("docgen", "datatest-stable"),
+    ("errmapgen", "diem-types"),
+    ("errmapgen", "datatest-stable"),
+    ("move-vm-natives", "diem-crypto"),
+    ("move-vm-runtime", "diem-state-view"),
+    ("move-vm-runtime", "diem-logger"),
+    ("move-vm-runtime", "diem-infallible"),
+    ("move-vm-runtime", "diem-crypto"),
+    ("move-vm-types", "diem-crypto"),
+    ("module-generation", "diem-types"),
+    ("test-generation", "language-e2e-tests"),
+    ("test-generation", "diem-vm"),
+    ("test-generation", "diem-types"),
+    ("test-generation", "diem-state-view"),
+    ("test-generation", "diem-logger"),
+    ("test-generation", "diem-config"),
+    ("disassembler", "diem-types"),
+    ("move-cli", "vm-genesis"),
+    ("move-cli", "diem-vm"),
+    ("move-cli", "diem-types"),
+    ("move-cli", "diem-state-view"),
+    ("move-cli", "diem-framework"),
+    ("move-cli", "datatest-stable"),
+    ("move-coverage", "diem-types"),
+    ("resource-viewer", "diem-types"),
+    ("resource-viewer", "diem-state-view"),
+    ("vm", "num-variants"),
+    ("vm", "diem-proptest-helpers"),
+    ("vm", "diem-crypto"),
+];
+
+const DIEM_CRATES_IN_LANGUAGE: &[&str] = &[
+    "diem-framework",
+    "diem-events-fetcher",
+    "diem-validator-interface",
+    "diem-transaction-replay",
+    "diem-writeset-generator",
+    "diem-vm",
+    "language-e2e-testsuite",
+    "ir-testsuite",
+    "language-e2e-tests",
+    "functional-tests",
+    "genesis-viewer",
+    "vm-genesis",
+    "transaction-builder",
+    "transaction-builder-generator",
+];
+
+// Determine if a crate is considered a Move crate or Diem crate.
+//
+// Current criterion:
+//   1. All crates outside language are considered Diem crates.
+//   2. All crates inside language are considered Move crates, unless marked otherwise.
+fn is_move_crate(crate_path: &str, crate_name: &str) -> bool {
+    if crate_path.starts_with("language/") {
+        if DIEM_CRATES_IN_LANGUAGE.contains(&crate_name) {
+            return false;
+        }
+        return true;
+    }
+
+    false
+}
+
+fn is_existing_move_to_diem_dep(name1: &str, name2: &str) -> bool {
+    EXISTING_MOVE_TO_DIEM_DEPS
+        .iter()
+        .any(|(a, b)| *a == name1 && *b == name2)
+}
```

Comment:
This could be expensive, may the list could be changed to be a HashSet<(str, str)> instead?